paths_get.py

```diff
@@ -0,0 +1,168 @@
+from typing import Iterable, Tuple, Union, Literal, Sequence, Dict
+
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+from anndata._core.raw import Raw
+
+
+# TODO: allow sequences: ("obsm", "X_pca", [0, 1])
+AnnDataPath = Union[
+    Tuple[Literal["X"], str],  # TODO: Support this or do ("layer", None, str)?
+    Tuple[Literal["layer"], str, str],
+    Tuple[Literal["obs"], str],
+    Tuple[Literal["obsm"], str, Union[int, str]],
+    Tuple[Literal["obsp"], str, Literal[0, 1]],  # TODO: actually needs one more parameter (cell name)
+    Tuple[Literal["raw"], Literal["X"], str],  # TODO: or "raw/X" and "raw/obs"?
+    Tuple[Literal["raw"], Literal["obs"], str],
+]
+
+
+def obs_df(
+    adata: AnnData,
+    paths: Iterable[Union[str, AnnDataPath]] = (),
+    *,
+    gene_symbols: AnnDataPath = None,
+    # no use_raw, no layer.
+) -> pd.DataFrame:
+    paths = list(map(normalize_adpath, paths))
+    names = paths_to_names(paths)
+    columns = {n: get_obs_vector(adata, p, gene_symbols) for n, p in names.items()}
+    return pd.DataFrame(columns, adata.obs_names)
+
+
+def get_obs_vector(adata: Union[AnnData, Raw], path: Union[str, AnnDataPath], gene_symbols: Union[str, AnnDataPath]):
+    path = normalize_adpath(path)
+    gene_symbols = None if gene_symbols is None else normalize_adpath(gene_symbols)
+    
+    if path[0] == "layer":  # X is here after normalizing
+        layer = adata.X if path[1] == "X" else adata.layers[path[1]]
+        idx = adata.var_names if gene_symbols is None else adata.var[gene_symbols]
+        # TODO: can layers also be data frames and have to be handled as obsm?
+        return layer[:, idx == path[2]]
+    if path[0] == "obs":
+        return adata.obs[path[1]]
+    assert not isinstance(adata, Raw)
+    if path[0] == "obsm":
+        m = adata.obsm[path[1]]
+        return m[path[2]] if isinstance(m, pd.DataFrame) else m[:, path[2]]
+    if path[0] == "obsp":
+        raise NotImplementedError("No good idea how to specify this. 'obsp/neighbors/CellName/1 vs .../2?'")
+    if path[0] == "raw":
+        return get_obs_vector(adata.raw, path[1:], gene_symbols)
+    else:
+        raise ValueError(f"Unknown AnnData attribute {path[0]!r} in path {path!r}")
+
+
+SHORT_CODES = dict(
+    X=["layer", "X"],
+    l=["layer"],
+    o=["obs"],
+    m=["obsm"],
+    p=["obsp"],
+    rX=["raw", "X"],
+    ro=["raw", "obs"],
+)
+
+
+def normalize_adpath(path: Union[str, AnnDataPath]) -> AnnDataPath:
+    # path is shorthand: "obs/Foo" and "o/Foo"
+    if not isinstance(path, str):
+        return path
+    path = path.split("/")
```

Comment:
Watch out for this, since column names in dataframes can currently have a `"/"` in them.

Also, we don't stop dataframe indices from having slashes. We're only expecting two "/" unless the value is in `uns`, right?

Reply:
So if we use slashes we need actual parsing then? I mean things are pretty deterministic, so we could set the limit for splitting depending on the attribute. That would be super headachy when combined with fallbacks and shorthand though.

Reply:
If I'm thinking about the same thing, I think we always needed actual parsing. Dataframe indices and columns can be arbitrary strings.